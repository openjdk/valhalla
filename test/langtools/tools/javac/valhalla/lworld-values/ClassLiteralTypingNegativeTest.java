/*
 * @test /nodynamiccopyright/
 * @bug 8221323
 * @summary  Javac should support class literals for projection types.
 * @compile/fail/ref=ClassLiteralTypingNegativeTest.out -XDrawDiagnostics ClassLiteralTypingNegativeTest.java
 */

public class ClassLiteralTypingNegativeTest {

    public static inline class Foo {
        final int value = 0;

        public static void main(String[] args) {
            Class<? extends Foo.ref> cFooRef = Foo.class; // Error
            cFooRef = new Foo().getClass(); // Error
            cFooRef = Foo.ref.class; // OK.
            cFooRef = Foo.val.class; // Error.
            Foo.val xv = new Foo();
            cFooRef = xv.getClass(); // Error
            Foo.ref xr = new Foo();
            cFooRef = xr.getClass(); // OK.
        }
    }

    interface I {}

    public static inline class Bar implements I {
        final int value = 0;

        public static void main(String[] args) {
            Class<? extends Bar.ref> cBarRef = Bar.class; // Error
            cBarRef = new Bar().getClass(); // Error
            cBarRef = Bar.ref.class; // OK.
            cBarRef = Bar.val.class; // Error.
            Bar.val xv = new Bar();
            cBarRef = xv.getClass(); // Error
            Bar.ref xr = new Bar();
            cBarRef = xr.getClass(); // OK.
        }
    }
}
