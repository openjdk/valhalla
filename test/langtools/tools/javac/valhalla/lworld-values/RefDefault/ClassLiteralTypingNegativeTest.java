/*
 * @test /nodynamiccopyright/
 * @bug 8221323
 * @summary  Javac should support class literals for projection types.
 * @compile/fail/ref=ClassLiteralTypingNegativeTest.out -XDrawDiagnostics ClassLiteralTypingNegativeTest.java
 */

public class ClassLiteralTypingNegativeTest {

    public static primitive class Foo.val {
        final int value = 0;

        public static void main(String[] args) {
            Class<? extends Foo> cFooRef = Foo.val.class; // Error
            cFooRef = ((Foo.val) new Foo()).getClass(); // OK
            cFooRef = Foo.class; // OK.
            cFooRef = Foo.val.class; // Error.
            Foo.val xv = new Foo();
            cFooRef = xv.getClass(); // OK
            Foo xr = new Foo();
            cFooRef = xr.getClass(); // OK.
        }
    }

    interface I {}

    public static primitive class Bar.val implements I {
        final int value = 0;

        public static void main(String[] args) {
            Class<? extends Bar> cBarRef = Bar.val.class; // Error
            cBarRef = ((Bar.val) new Bar()).getClass(); // OK
            cBarRef = Bar.class; // OK.
            cBarRef = Bar.val.class; // Error.
            Bar.val xv = new Bar();
            cBarRef = xv.getClass(); // OK.
            Bar xr = new Bar();
            cBarRef = xr.getClass(); // OK.
        }
    }
}
