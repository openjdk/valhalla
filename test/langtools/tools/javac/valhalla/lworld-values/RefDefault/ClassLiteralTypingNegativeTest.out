ClassLiteralTypingNegativeTest.java:14:51: compiler.err.prob.found.req: (compiler.misc.inconvertible.types: java.lang.Class<compiler.misc.type.captureof: 1, ? extends java.lang.Object>, java.lang.Class<? extends ClassLiteralTypingNegativeTest.Foo>)
ClassLiteralTypingNegativeTest.java:17:30: compiler.err.prob.found.req: (compiler.misc.inconvertible.types: java.lang.Class<compiler.misc.type.captureof: 1, ? extends java.lang.Object>, java.lang.Class<? extends ClassLiteralTypingNegativeTest.Foo>)
ClassLiteralTypingNegativeTest.java:31:51: compiler.err.prob.found.req: (compiler.misc.inconvertible.types: java.lang.Class<compiler.misc.type.captureof: 1, ? extends java.lang.Object&ClassLiteralTypingNegativeTest.I>, java.lang.Class<? extends ClassLiteralTypingNegativeTest.Bar>)
ClassLiteralTypingNegativeTest.java:34:30: compiler.err.prob.found.req: (compiler.misc.inconvertible.types: java.lang.Class<compiler.misc.type.captureof: 1, ? extends java.lang.Object&ClassLiteralTypingNegativeTest.I>, java.lang.Class<? extends ClassLiteralTypingNegativeTest.Bar>)
4 errors
